File: bindings/cpp/examples/robstride_wrapper_demo.cpp
#include <chrono>
#include <iostream>
#include <string>
#include <thread>

#include "motorbridge/motorbridge.hpp"

int main(int argc, char** argv) {
  try {
    std::string channel = "can0";
    std::string model = "rs-00";
    uint16_t motor_id = 127;
    uint16_t feedback_id = 0xFF;
    std::string mode = "ping";

    if (argc > 1) channel = argv[1];
    if (argc > 2) model = argv[2];
    if (argc > 3) motor_id = static_cast<uint16_t>(std::stoul(argv[3], nullptr, 0));
    if (argc > 4) feedback_id = static_cast<uint16_t>(std::stoul(argv[4], nullptr, 0));
    if (argc > 5) mode = argv[5];

    motorbridge::Controller ctrl(channel);
    auto motor = ctrl.add_robstride_motor(motor_id, feedback_id, model);

    if (mode == "ping") {
      auto [device_id, responder_id] = motor.robstride_ping();
      std::cout << "ping ok device_id=" << static_cast<int>(device_id)
                << " responder_id=" << static_cast<int>(responder_id) << "\n";
      auto st = motor.get_state();
      if (st.has_value()) std::cout << "state pos=" << st->pos << " vel=" << st->vel << "\n";
      ctrl.shutdown();
      return 0;
    }

    if (mode == "read-param") {
      float mech_pos = motor.robstride_get_param_f32(0x7019);
      float vbus = motor.robstride_get_param_f32(0x701C);
      std::cout << "mechanical_position=" << mech_pos << " vbus=" << vbus << "\n";
      ctrl.shutdown();
      return 0;
    }

    ctrl.enable_all();
    motor.ensure_mode(mode == "mit" ? motorbridge::Mode::MIT : motorbridge::Mode::VEL, 1000);

    for (int i = 0; i < 20; ++i) {
      if (mode == "mit") {
        motor.send_mit(0.0f, 0.0f, 8.0f, 0.2f, 0.0f);
      } else {
        motor.send_vel(0.2f);
      }
      auto st = motor.get_state();
      if (st.has_value()) {
        std::cout << "#" << i << " pos=" << st->pos << " vel=" << st->vel << " torq=" << st->torq
                  << " status=" << static_cast<int>(st->status_code) << "\n";
      } else {
        std::cout << "#" << i << " no feedback yet\n";
      }
      std::this_thread::sleep_for(std::chrono::milliseconds(50));
    }

    ctrl.shutdown();
    return 0;
  } catch (const std::exception& e) {
    std::cerr << "error: " << e.what() << "\n";
    return 1;
  }
}

File: bindings/python/examples/robstride_wrapper_demo.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import time

from motorbridge import Controller, Mode


def _parse_id(text: str) -> int:
    return int(text, 0)


def main() -> None:
    parser = argparse.ArgumentParser(description="RobStride wrapper demo for Python SDK")
    parser.add_argument("--channel", default="can0")
    parser.add_argument("--model", default="rs-00")
    parser.add_argument("--motor-id", default="127")
    parser.add_argument("--feedback-id", default="0xFF")
    parser.add_argument("--mode", choices=["ping", "read-param", "mit", "vel"], default="ping")
    parser.add_argument("--pos", type=float, default=0.0)
    parser.add_argument("--vel", type=float, default=0.0)
    parser.add_argument("--kp", type=float, default=8.0)
    parser.add_argument("--kd", type=float, default=0.2)
    parser.add_argument("--tau", type=float, default=0.0)
    parser.add_argument("--loop", type=int, default=20)
    parser.add_argument("--dt-ms", type=int, default=50)
    parser.add_argument("--param-id", default="0x7019")
    parser.add_argument("--param-timeout-ms", type=int, default=1000)
    args = parser.parse_args()

    motor_id = _parse_id(args.motor_id)
    feedback_id = _parse_id(args.feedback_id)
    param_id = _parse_id(args.param_id)

    with Controller(args.channel) as ctrl:
        motor = ctrl.add_robstride_motor(motor_id, feedback_id, args.model)
        try:
            if args.mode == "ping":
                device_id, responder_id = motor.robstride_ping()
                print(f"ping ok device_id={device_id} responder_id={responder_id}")
                print(motor.get_state())
                return

            if args.mode == "read-param":
                value = motor.robstride_get_param_f32(param_id, args.param_timeout_ms)
                print(f"param 0x{param_id:04X} = {value}")
                print(motor.get_state())
                return

            ctrl.enable_all()
            motor.ensure_mode(Mode.MIT if args.mode == "mit" else Mode.VEL, 1000)

            for i in range(args.loop):
                if args.mode == "mit":
                    motor.send_mit(args.pos, args.vel, args.kp, args.kd, args.tau)
                else:
                    motor.send_vel(args.vel)
                print(f"#{i} {motor.get_state()}")
                if args.dt_ms > 0:
                    time.sleep(args.dt_ms / 1000.0)
        finally:
            motor.close()


if __name__ == "__main__":
    main()
